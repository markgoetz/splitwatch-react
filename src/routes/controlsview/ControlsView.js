import React, { Component } from 'react';
import Header from '../../components/Header/Header';
import Button from '../../components/Button/Button';
import VerticalLayout from '../../components/VerticalLayout/VerticalLayout';
import HorizontalLayout from '../../components/HorizontalLayout/HorizontalLayout';
import Wrapper from '../../components/Wrapper/Wrapper';
import TimerProvider from '../../components/TimerProvider/TimerProvider';
import TimerComponent from '../../components/TimerComponent/TimerComponent';
import LastSplit from '../../components/LastSplit/LastSplit';
import SplitList from '../../components/SplitList/SplitList';
import StartSplitButton from '../../components/StartSplitButton/StartSplitButton';
import SidebarLayout from '../../components/SidebarLayout/SidebarLayout';
import TimerContext from '../../lib/TimerContext';

export default class ControlsView extends Component {
    componentWillReceiveProps(nextProps) {
        if (nextProps.baseTime !== this.props.baseTime) {
            this.timer.setBaseTime(nextProps.baseTime);
        }
    }

    render() {

        return (<div>
            <Header />
            <TimerProvider>
                <Wrapper align="center">
                    <VerticalLayout align="center">
                        <TimerComponent />
                        <LastSplit />
                        <SidebarLayout>
                            <VerticalLayout>
                                <StartSplitButton />
                                <HorizontalLayout>
                                    <TimerContext.Consumer>
                                        {(context => <>
                                            <Button onClick={() => context.pause(Date.now())}>Pause</Button>
                                            <Button onClick={() => context.stop(Date.now())}>Stop</Button>
                                            <Button onClick={() => context.reset(Date.now())}>Reset</Button>
                                        </> )}
                                    </TimerContext.Consumer>
                                </HorizontalLayout>
                            </VerticalLayout>
                            <SplitList />
                        </SidebarLayout>
                    </VerticalLayout>
                </Wrapper>
            </TimerProvider>
        </div>);
    }
}
