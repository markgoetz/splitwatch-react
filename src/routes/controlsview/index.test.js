import { h } from 'preact';
import ControlsView from './index';
import Timer from '../../lib/Timer';
jest.mock('../../lib/Timer.js');
Timer.mockImplementation();

import { shallow } from 'preact-render-spy';
const testNow = new Date('2018-01-01');
let component;

beforeEach(() => {
  Timer.mockReset();
  require('jest-mock-now')(testNow);

  component = new ControlsView();
  component.props = {
    startTime: 1000,
    baseTime: 3000
  };
  component.componentDidMount();
});

describe('ControlsView', () => {
  test('renders when playing', () => {
    const context = shallow(<ControlsView isPlaying startTime={1000} baseTime={3000} splits={[]} />);
    expect(context).toMatchSnapshot();
  });

  test('renders when paused', () => {
    const context = shallow(<ControlsView startTime={1000} baseTime={3000} splits={[]} />);
    expect(context).toMatchSnapshot();
  });

  test('renders with splits', () => {
    const context = shallow(<ControlsView startTime={1000} baseTime={3000} splits={[100, 200, 300]} />);
    expect(context).toMatchSnapshot();
  });

  test('creates Timer object on instantiation', () => {
    expect(component.timer).toBeInstanceOf(Timer);
    expect(Timer.mock.calls[0]).toEqual([1000, 3000]);
  });

  test('updates the timer basetime when its props change', () => {
    component.timer.setBaseTime.mockImplementation();

    component.componentWillReceiveProps({ baseTime: 5000 });
    expect(component.timer.setBaseTime).toHaveBeenCalledWith(5000);
  });

  test('does not update the timer basetime when its props does change', () => {
    component.timer.setBaseTime.mockImplementation();

    component.componentWillReceiveProps({ baseTime: 3000 });
    expect(component.timer.setBaseTime).toHaveBeenCalledTimes(0);
  });

  test('calls Timer start when starts', () => {
    component.timer.start.mockImplementation();
    component.startTime();
    expect(component.timer.start).toHaveBeenCalledWith(testNow.getTime());
  });

  test('calls Timer pause when pauses', () => {
    component.timer.pause.mockImplementation();
    component.pauseTime();
    expect(component.timer.pause).toHaveBeenCalledWith(testNow.getTime());
  });
  
  test('calls Timer stop when stops', () => {
    component.timer.stop.mockImplementation();
    component.stopTime();
    expect(component.timer.stop).toHaveBeenCalledWith(testNow.getTime());
  });
  
  test('calls Timer reset when resets', () => {
    component.timer.reset.mockImplementation();
    component.resetTime();
    expect(component.timer.reset).toHaveBeenCalledWith(testNow.getTime());
  });

  test('calls Timer split when splits', () => {
    component.timer.reset.mockImplementation();
    component.addSplit();
    expect(component.timer.split).toHaveBeenCalledWith(testNow.getTime());
  });
});