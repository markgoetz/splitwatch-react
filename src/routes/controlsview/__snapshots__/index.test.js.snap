// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsView renders when paused 1`] = `
preact-render-spy (1 nodes)
-------
<div>
  <Header></Header>
  <Wrapper align="center">
    <VerticalLayout align="center">
      <TimerComponent
        startTime={1000}
        baseTime={3000}
        splits={Array []}
      >
      </TimerComponent>
      <LastSplit></LastSplit>
      <SidebarLayout>
        <VerticalLayout>
          <Button
            isPrimary={true}
            onClick={[Function anonymous]}
          >
            Start
          </Button>
          <HorizontalLayout>
            <Button onClick={[Function anonymous]}>Pause</Button>
            <Button onClick={[Function anonymous]}>Stop</Button>
            <Button onClick={[Function anonymous]}>Reset</Button>
          </HorizontalLayout>
        </VerticalLayout>
        <SplitList splits={Array []}></SplitList>
      </SidebarLayout>
    </VerticalLayout>
  </Wrapper>
</div>

`;

exports[`ControlsView renders when playing 1`] = `
preact-render-spy (1 nodes)
-------
<div>
  <Header></Header>
  <Wrapper align="center">
    <VerticalLayout align="center">
      <TimerComponent
        isPlaying={true}
        startTime={1000}
        baseTime={3000}
        splits={Array []}
      >
      </TimerComponent>
      <LastSplit></LastSplit>
      <SidebarLayout>
        <VerticalLayout>
          <Button
            isPrimary={true}
            onClick={[Function anonymous]}
          >
            Split
          </Button>
          <HorizontalLayout>
            <Button onClick={[Function anonymous]}>Pause</Button>
            <Button onClick={[Function anonymous]}>Stop</Button>
            <Button onClick={[Function anonymous]}>Reset</Button>
          </HorizontalLayout>
        </VerticalLayout>
        <SplitList splits={Array []}></SplitList>
      </SidebarLayout>
    </VerticalLayout>
  </Wrapper>
</div>

`;

exports[`ControlsView renders with splits 1`] = `
preact-render-spy (1 nodes)
-------
<div>
  <Header></Header>
  <Wrapper align="center">
    <VerticalLayout align="center">
      <TimerComponent
        startTime={1000}
        baseTime={3000}
        splits={
          Array [
            100,
            200,
            300
          ]
        }
      >
      </TimerComponent>
      <LastSplit split={300}></LastSplit>
      <SidebarLayout>
        <VerticalLayout>
          <Button
            isPrimary={true}
            onClick={[Function anonymous]}
          >
            Start
          </Button>
          <HorizontalLayout>
            <Button onClick={[Function anonymous]}>Pause</Button>
            <Button onClick={[Function anonymous]}>Stop</Button>
            <Button onClick={[Function anonymous]}>Reset</Button>
          </HorizontalLayout>
        </VerticalLayout>
        <SplitList
          splits={
            Array [
              100,
              200,
              300
            ]
          }
        >
        </SplitList>
      </SidebarLayout>
    </VerticalLayout>
  </Wrapper>
</div>

`;
